### ISSUE TITLE
Multiple predicted viral regions in similar area of contig.

### ISSUE THREAD

**Issue by genomics-pixel** [simroux/VirSorter#48]
Dear Simon,

Thank you for your help with the earlier issue.

Though it seemed that the problems were solved, one issue unfortunately persisted.
(I made a new issue as this may be a different problem and because the last one was quite long.)

When I have run VirSorter/1.0.5 on some of my other assemblies I noticed that
multiple viral regions were predicted on nearly identical area of input contig.

Ex.)
\>VIRSorter_k251_224725_flag_0_multi_37_0000_len_22303_ID_SAMPLE2_COUNT_3641_**gene_9_gene_20**-**8464-21493**-cat_4
\>VIRSorter_k251_224725_flag_0_multi_37_0000_len_22303_ID_SAMPLE2_COUNT_3641_**gene_9_gene_18**-**8464-18107**-cat_4

I have tried using blastall and also the following commands with diamond and the "--no_c" option, but the same thing appeared.
  - wrapper_phage_contigs_sorter_iPlant.pl **-f** [input contig fasta file] **--db 2** **-wdir** [output contig fasta file]
  - wrapper_phage_contigs_sorter_iPlant.pl **-f** [input contig fasta file] **--db 2** **--diamond** **-wdir** [output contig fasta file]
  - wrapper_phage_contigs_sorter_iPlant.pl **-f** [input contig fasta file] **--db 2** **--diamond** **--no_c** **-wdir** [output contig fasta file]

In other cases I have seen nearly identical pair of  predicted viral regions predicted in different categorized (ex. 4 and 5).

Is this the expected behavior of VirSorter or am I getting something wrong? 

Sincerely,
genomics-pixel

**Edit:** I went through your paper on VirSorter published in PeerJ and came across "Sequence metrics summary" paragraph, which states that VirSorter merges overlapping predictions. I am not sure, but perhaps the current problem is related with this part of the prediction step?

**Comment by genomics-pixel** [comment]
Dear Simon,

My apologies for sending another message in succession.
I took a further look into the analysis results and summarized what I have found.

Before getting into the details, the following command was executed.
(the input file is the _same_ as the one used in above post but with _shortened fasta header_ for simplicity)
* **wrapper_phage_contigs_sorter_iPlant.pl** -f [input contig fasta file] **--db 2** **--diamond** -wdir [output contig fasta file]
* The log files ("logs/[out"](https://github.com/simroux/VirSorter/files/3573966/out.txt.tar.gz), "logs/[err"](https://github.com/simroux/VirSorter/files/3573969/err.txt))  and the "Metrics_files/[VIRSorter_affi-contigs.tab](https://github.com/simroux/VirSorter/files/3573807/VIRSorter_affi-contigs.tab.tar.gz)" file seems to be fine, so I believe that hmmer and DIAMOND searches have performed smoothly without problem.

As mentioned in the above post, I noticed that there were multiple predicted viral regions in similar area of the same contig. In the case of **"k251z78343" contig**, there were **two very similar regions: "gene 93 to gene 229" and "gene 93 to gene 448"**:
\>VIRSorter_**k251z78343_gene_93_gene_248**-71397-207802-cat_5
\>VIRSorter_**k251z78343_gene_93_gene_229**-71397-188491-cat_5

You can also see that each predictions have treated this annotation-sparse region differently by looking into the "[VIRSorter_global-phage-signal.csv](https://github.com/simroux/VirSorter/files/3573842/VIRSorter_global-phage-signal.csv.txt)" file (i.e., "gene_149-gene_248" region is picked up only in the second longer prediction).

The following is a subset of "VIRSorter_global-phage-signal.csv" (Perhaps better to copy&paste to excel for better visibility.)
>\## Contig_id,Nb genes contigs,Fragment,Nb genes,Category,Nb phage hallmark genes,Phage gene enrichment sig,Non-Caudovirales phage gene enrichment sig,Pfam depletion sig,Uncharacterized enrichment sig,Strand switch depletion sig,Short genes enrichment sig
**VIRSorter_k251z78343**,260,VIRSorter_k251z78343-gene_93-gene_229,137,2,,gene_130-gene_229:4.87224408777739,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890
**VIRSorter_k251z78343**,260,VIRSorter_k251z78343-gene_93-gene_248,156,2,,gene_130-gene_229:6.86349848266885;_**gene_149-gene_248:2.97349330748608**_,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890,gene_93-gene_192:50.00000351237890

In the "VIRSorter_k251z78343" rows/entries of "Metrics_files/[VIRSorter_affi-contigs.tab](https://github.com/simroux/VirSorter/files/3573807/VIRSorter_affi-contigs.tab.tar.gz)",
I noticed that there was a region with _very few annotated genes_ (only 2 annotated genes) in between **gene_229 and gene_241** which could be the source of difference in the prediction.

In the above analysis files, you can also see that something quite similar also happened to contig id: "k251z224725".

As a summary, my best guess is that for some reason, same genes listed in "Metrics_files/VIRSorter_affi-contgs.tab" are used to predict viral sequence _more than once_  causing multiple viral sequences which are nearly identical to be placed in "VIRSorter_global-phage-signal.csv".
This seems a bit strange to me, because this problem only occurs for one or two contigs while all other contigs do not have overlapping predicted viral sequences. 

I am terribly sorry to have to ask you for your help again.
VirSorter is a fantastic software and I would really appreciate it if you could assist me in solving this matter.

Thank you in advance.

Sincerely,
genomics-pixel

P.S.
If you need the entire output directory of the above directory, please tell me.

**Comment by simroux** [comment]
Hi,

I agree that this looks like bug: could you send a fasta file with these two contigs so I can try to reproduce it ? I don't see anything obvious that would explain this right now unfortunately.

Best,
Simon

**Comment by genomics-pixel** [comment]
Dear Simon,

Thank you for taking the time to look into this problem.

I have sent you an e-mail with the link and password to download the
entire output directory of VirSorter (1.0.5).

The command ran to obtain this output is the following:
**wrapper_phage_contigs_sorter_iPlant.pl** **-f** [input contig fasta file] **--db 2 --diamond -wdir** [output contig fasta file] **--ncpu 4**

If you have not recieved my e-mail or if you have any questions please let me know.
Again, I deeply appreciate your help.

Sincerely,
genomics-pixel

**Comment by simroux** [comment]
Hi ! 
There was a bug in 1.0.5 with some prophage predictions that your test case uncovered. I just pushed a new version of the script that should have solved it, so in theory if you pull this latest version and run it on the same file  you should see:
- only 1 prophage predicted for k251z78343
- only an r_0 and r_1 folder in VirSorter output directory, not r_0 to r_11 (this is another small bug I found and fixed).

Let me know if this new version didn't fix things on your side, or if you find anything else !

Best,
Simon

**Comment by genomics-pixel** [comment]
Dear Simon,

Great to know that the bug is fixed!

I will have the system administrator install the updated version of VirSorter
on our computer cluster and test it on my side as well.

Once the test is finished I will get back to you as soon as possible.

Many thanks for your kind help.

Sincerely,
genomics-pixel

**Comment by genomics-pixel** [comment]
Dear Simon,

Sorry for my reply being late.
I had a high fever and could not work for a while.

I used your new version of VirSorter to test my data and the results looked good! 
(i.e. only one predicted viral region for  k251z78343 and only an r_0 and r_1 folder 
in VirSorter output directory).

I will now try to use it on the rest of my assemblies, but I have one question:
Would it make sense to use VirSorter to analyse assemblies from each samples 
**individually** instead of using VirSorter to analyze a single fasta which contains
all assemblies?

I have more than one thousand assemblies from different samples along with 
more than a hundred co-assemblies and its quite difficult to process them all
at once.

Sincerely,
genomics-pixel

**Comment by simroux** [comment]
Hi,
It shouldn't matter much whether you run VirSorter on individual assemblies or on a pool of all contigs, so for simplicity (and speed), I would suggest processing them separately as you can run all of them in parallel.

Best,
Simon

**Comment by genomics-pixel** [comment]
Dear Simon,

Thank you for your helpful advice!
I am glad that there is no need to pool the assemblies together.

Again many thanks!

Sincerely,
genomics-pixel
